internal/singleflight/singleflight/singleflight_test.go

```diff
@@ -0,0 +1,273 @@
+						return res, nil
+					},
+				},
+				want: want{
+					wantV:      res,
+					wantShared: false,
+					err:        nil,
+				},
+				beforeFunc: func(g Group, args args) {
+					gcnt := 10
+					ch := make(chan struct{}, gcnt)
+
+					for i := 0; i < gcnt; i++ {
+						wg.Add(1)
+						go func(i int) {
+							ch <- struct{}{}
+							defer wg.Done()
+							condWaitFn()
+
+							_, _, _ = g.Do(context.Background(), strconv.Itoa(i), func() (interface{}, error) {
+								time.Sleep(time.Nanosecond * 100)
+								atomic.AddUint32(&cnt, 1)
+								return "vdaas/vald", nil
+							})
+						}(i)
+					}
+
+					for i := 0; i < gcnt; i++ {
+						<-ch
+					}
+					close(ch)
+				},
+				checkFunc: func(w want, gotV interface{}, gotShared bool, err error) error {
+					if got, want := int(atomic.LoadUint32(&cnt)), 11; got != want {
+						return errors.Errorf("cnt got = %d, want = %d", got, want)
+					}
+
+					if err := defaultCheckFunc(w, gotV, gotShared, err); err != nil {
+						return err
+					}
+					return nil
+				},
+			}
+		}(),
+
+		func() test {
+			var cnt uint32
+			var res string = "res_1"
+
+			var (
+				mu         = new(sync.Mutex)
+				cond       = sync.NewCond(mu)
+				wg         = new(sync.WaitGroup)
+				condWaitFn = func() {
+					mu.Lock()
+					defer mu.Unlock()
+					cond.Wait()
+				}
+			)
+
+			return test{
+				name: "returns (v, shared, nil) when Do is called with same key",
+				args: args{
+					key: "req_1",
+					ctx: context.Background(),
+					fn: func() (interface{}, error) {
+						atomic.AddUint32(&cnt, 1)
+						return res, nil
+					},
+				},
+				fields: fields{
+					m: make(map[string]*call),
+				},
+				util: util{
+					mu:         mu,
+					cond:       cond,
+					wg:         wg,
+					condWaitFn: condWaitFn,
+				},
+				want: want{
+					wantV:      res,
+					wantShared: true,
+					err:        nil,
+				},
+				beforeFunc: func(g Group, args args) {
+					wg.Add(1)
+					go func() {
+						defer wg.Done()
+						_, _, _ = g.Do(context.Background(), args.key, func() (interface{}, error) {
+							time.Sleep(3 * time.Second)
+							return args.fn()
+						})
+					}()
+
+					gcnt := 10
+					ch := make(chan struct{}, gcnt)
+
+					for i := 0; i < gcnt; i++ {
+						wg.Add(1)
+						go func() {
+							ch <- struct{}{}
+							defer wg.Done()
+							condWaitFn()
+
+							_, _, _ = g.Do(context.Background(), args.key, func() (interface{}, error) {
+								atomic.AddUint32(&cnt, 1)
+								return "vdaas/vald", nil
+							})
+						}()
+					}
+
+					for i := 0; i < gcnt; i++ {
+						<-ch
+					}
+					close(ch)
+				},
+				checkFunc: func(w want, gotV interface{}, gotShared bool, err error) error {
+					if got, want := int(atomic.LoadUint32(&cnt)), 1; got != want {
+						return errors.Errorf("cnt got = %d, want = %d", got, want)
+					}
+
+					if err := defaultCheckFunc(w, gotV, gotShared, err); err != nil {
+						return err
+					}
+					return nil
+				},
+			}
+		}(),
```

Comment:
please add fail test case that `fn()` returns error and do not cache the result.

Reply:
wait... it should cache the fail result 😂